File: FirstBadVersionSolution.java
// 15:30 to 15:42

/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class FirstBadVersionSolution extends VersionControl {
    public int firstBadVersion(int n) {
        int left = 1;
        int right = n;
        int mid;
        int ans =-1;
        while ( left<=right){
            
            mid =  left + ((right - left)/2 ); 
            
            if(isBadVersion(mid)){
                right = mid-1;
                ans=mid;
            }
            else {
                left = mid+1;
            }
                        
        }
        return ans;
    }
}

File: FirstBadVersionSolution.cs
/* Code for first bad verison in leet code but done recursively*/

public class Solution : VersionControl {
    
    public int recursiveSolution(int left, int right){
            
		/* check if right depases left at any point*/
            if(right>left){

                int mid = (right+left)/2;

                if(IsBadVersion(mid)==true){
                    return recursiveSolution(left,mid);
                }
                else{
                    return recursiveSolution(mid+1,right);
                }         
            }
            /* if it does that means that we only have one item left on our array so it must be the first bad version, we check before returning it tho by adding and if clause in the posible chance that the array has no bad versions*/
            else{
                return left;
            }
        }
    
    public int FirstBadVersion(int n) {
        
        int firstBad;
        
        firstBad= recursiveSolution(1,n);
        return firstBad;
        
    }
}